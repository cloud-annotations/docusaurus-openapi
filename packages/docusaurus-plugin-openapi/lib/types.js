"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Type = void 0;
var Type;
(function (Type) {
    Type["string"] = "string";
    Type["number"] = "number";
    Type["integer"] = "integer";
    Type["boolean"] = "boolean";
    Type["object"] = "object";
    Type["array"] = "array";
})(Type = exports.Type || (exports.Type = {}));
