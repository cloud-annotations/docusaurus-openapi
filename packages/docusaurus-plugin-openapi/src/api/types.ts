/* ============================================================================
 * Copyright (c) Cloud Annotations
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import type { JSONSchema4, JSONSchema6, JSONSchema7 } from "json-schema";

interface Map<T> {
  [key: string]: T;
}

export interface OpenApiObject {
  openapi: string;
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObject;
  components?: ComponentsObject;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

export interface OpenApiObjectWithRef {
  openapi: string;
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObjectWithRef;
  components?: ComponentsObjectWithRef;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

interface InfoObject {
  title: string;
  description?: string;
  termsOfService?: string;
  contact?: ContactObject;
  license?: LicenseObject;
  version: string;
}

interface ContactObject {
  name?: string;
  url?: string;
  email?: string;
}

interface LicenseObject {
  name: string;
  url?: string;
}

interface ServerObject {
  url: string;
  description?: string;
  variables?: Map<ServerVariable>;
}

interface ServerVariable {
  enum?: string[];
  default: string;
  description?: string;
}

interface ComponentsObject {
  schemas?: Map<SchemaObject>;
  responses?: Map<ResponseObject>;
  parameters?: Map<ParameterObject>;
  examples?: Map<ExampleObject>;
  requestBodies?: Map<RequestBodyObject>;
  headers?: Map<HeaderObject>;
  securitySchemes?: Map<SecuritySchemeObject>;
  links?: Map<LinkObject>;
  callbacks?: Map<CallbackObject>;
}

interface ComponentsObjectWithRef {
  schemas?: Map<SchemaObjectWithRef | ReferenceObject>;
  responses?: Map<ResponseObjectWithRef | ReferenceObject>;
  parameters?: Map<ParameterObjectWithRef | ReferenceObject>;
  examples?: Map<ExampleObject | ReferenceObject>;
  requestBodies?: Map<RequestBodyObjectWithRef | ReferenceObject>;
  headers?: Map<HeaderObjectWithRef | ReferenceObject>;
  securitySchemes?: Map<SecuritySchemeObject | ReferenceObject>;
  links?: Map<LinkObject | ReferenceObject>;
  callbacks?: Map<CallbackObjectWithRef | ReferenceObject>;
}

type PathsObject = Map<PathItemObject>;

type PathsObjectWithRef = Map<PathItemObjectWithRef>;

interface PathItemObject {
  $ref?: string;
  summary?: string;
  description?: string;
  get?: OperationObject;
  put?: OperationObject;
  post?: OperationObject;
  delete?: OperationObject;
  options?: OperationObject;
  head?: OperationObject;
  patch?: OperationObject;
  trace?: OperationObject;
  servers?: ServerObject[];
  parameters?: ParameterObject[];
}

interface PathItemObjectWithRef {
  $ref?: string;
  summary?: string;
  description?: string;
  get?: OperationObjectWithRef;
  put?: OperationObjectWithRef;
  post?: OperationObjectWithRef;
  delete?: OperationObjectWithRef;
  options?: OperationObjectWithRef;
  head?: OperationObjectWithRef;
  patch?: OperationObjectWithRef;
  trace?: OperationObjectWithRef;
  servers?: ServerObject[];
  parameters?: (ParameterObjectWithRef | ReferenceObject)[];
}

interface OperationObject {
  tags?: string[];
  summary?: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
  operationId?: string;
  parameters?: ParameterObject[];
  requestBody?: RequestBodyObject;
  responses: ResponsesObject;
  callbacks?: Map<CallbackObject>;
  deprecated?: boolean;
  security?: SecurityRequirementObject[];
  servers?: ServerObject[];

  // extensions
  "x-deprecated-description"?: string;
}

interface OperationObjectWithRef {
  tags?: string[];
  summary?: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
  operationId?: string;
  parameters?: (ParameterObjectWithRef | ReferenceObject)[];
  requestBody?: RequestBodyObjectWithRef | ReferenceObject;
  responses: ResponsesObjectWithRef;
  callbacks?: Map<CallbackObjectWithRef | ReferenceObject>;
  deprecated?: boolean;
  security?: SecurityRequirementObject[];
  servers?: ServerObject[];

  // extensions
  "x-deprecated-description"?: string;
}

interface ExternalDocumentationObject {
  description?: string;
  url: string;
}

interface ParameterObject {
  name: string;
  in: string;
  description?: string;
  required?: boolean;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  //
  style?: string;
  explode?: string;
  allowReserved?: boolean;
  schema?: SchemaObject;
  example?: any;
  examples?: Map<ExampleObject>;
  //
  content?: Map<MediaTypeObject>;
  // ignoring stylings: matrix, label, form, simple, spaceDelimited,
  // pipeDelimited and deepObject
}

interface ParameterObjectWithRef {
  name: string;
  in: string;
  description?: string;
  required?: boolean;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  //
  style?: string;
  explode?: string;
  allowReserved?: boolean;
  schema?: SchemaObjectWithRef | ReferenceObject;
  example?: any;
  examples?: Map<ExampleObject | ReferenceObject>;
  //
  content?: Map<MediaTypeObjectWithRef>;
  // ignoring stylings: matrix, label, form, simple, spaceDelimited,
  // pipeDelimited and deepObject
}

interface RequestBodyObject {
  description?: string;
  content: Map<MediaTypeObject>;
  required?: boolean;
}

interface RequestBodyObjectWithRef {
  description?: string;
  content: Map<MediaTypeObjectWithRef>;
  required?: boolean;
}

interface MediaTypeObject {
  schema?: SchemaObject;
  example?: any;
  examples?: Map<ExampleObject>;
  encoding?: Map<EncodingObject>;
}

interface MediaTypeObjectWithRef {
  schema?: SchemaObjectWithRef | ReferenceObject;
  example?: any;
  examples?: Map<ExampleObject | ReferenceObject>;
  encoding?: Map<EncodingObjectWithRef>;
}

interface EncodingObject {
  contentType?: string;
  headers?: Map<HeaderObject>;
  style?: string;
  explode?: boolean;
  allowReserved?: boolean;
}

interface EncodingObjectWithRef {
  contentType?: string;
  headers?: Map<HeaderObjectWithRef | ReferenceObject>;
  style?: string;
  explode?: boolean;
  allowReserved?: boolean;
}

type ResponsesObject = Map<ResponseObject>;

type ResponsesObjectWithRef = Map<ResponseObjectWithRef | ReferenceObject>;

interface ResponseObject {
  description: string;
  headers?: Map<HeaderObject>;
  content?: Map<MediaTypeObject>;
  links?: Map<LinkObject>;
}

interface ResponseObjectWithRef {
  description: string;
  headers?: Map<HeaderObjectWithRef | ReferenceObject>;
  content?: Map<MediaTypeObjectWithRef>;
  links?: Map<LinkObject | ReferenceObject>;
}

type CallbackObject = Map<PathItemObject>;

type CallbackObjectWithRef = Map<PathItemObjectWithRef>;

interface ExampleObject {
  summary?: string;
  description?: string;
  value?: any;
  externalValue?: string;
}

interface LinkObject {
  operationRef?: string;
  operationId?: string;
  parameters?: Map<any>;
  requestBody?: any;
  description?: string;
  server?: ServerObject;
}

type HeaderObject = Omit<ParameterObject, "name" | "in">;

type HeaderObjectWithRef = Omit<ParameterObjectWithRef, "name" | "in">;

interface TagObject {
  name: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
}

interface ReferenceObject {
  $ref: string;
}

type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;
type SchemaObject = Omit<
  JSONSchema,
  | "type"
  | "allOf"
  | "oneOf"
  | "anyOf"
  | "not"
  | "items"
  | "properties"
  | "additionalProperties"
> & {
  // OpenAPI specific overrides
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array";
  allOf?: SchemaObject;
  oneOf?: SchemaObject;
  anyOf?: SchemaObject;
  not?: SchemaObject;
  items?: SchemaObject;
  properties?: SchemaObject;
  additionalProperties?: boolean | SchemaObject;

  // OpenAPI additions
  nullable?: boolean;
  discriminator?: DiscriminatorObject;
  readOnly?: boolean;
  writeOnly?: boolean;
  xml?: XMLObject;
  externalDocs?: ExternalDocumentationObject;
  example?: any;
  deprecated?: boolean;
};

type SchemaObjectWithRef = Omit<
  JSONSchema,
  | "type"
  | "allOf"
  | "oneOf"
  | "anyOf"
  | "not"
  | "items"
  | "properties"
  | "additionalProperties"
> & {
  // OpenAPI specific overrides
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array";
  allOf?: SchemaObject | ReferenceObject;
  oneOf?: SchemaObject | ReferenceObject;
  anyOf?: SchemaObject | ReferenceObject;
  not?: SchemaObject | ReferenceObject;
  items?: SchemaObject | ReferenceObject;
  properties?: SchemaObject | ReferenceObject;
  additionalProperties?: boolean | SchemaObject | ReferenceObject;

  // OpenAPI additions
  nullable?: boolean;
  discriminator?: DiscriminatorObject;
  readOnly?: boolean;
  writeOnly?: boolean;
  xml?: XMLObject;
  externalDocs?: ExternalDocumentationObject;
  example?: any;
  deprecated?: boolean;
};

interface DiscriminatorObject {
  propertyName: string;
  mapping?: Map<string>;
}

interface XMLObject {
  name?: string;
  namespace?: string;
  prefix?: string;
  attribute?: boolean;
  wrapped?: boolean;
}

type SecuritySchemeObject =
  | ApiKeySecuritySchemeObject
  | HttpSecuritySchemeObject
  | Oauth2SecuritySchemeObject
  | OpenIdConnectSecuritySchemeObject;

interface ApiKeySecuritySchemeObject {
  type: "apiKey";
  description?: string;
  name: string;
  in: "query" | "header" | "cookie";
}

interface HttpSecuritySchemeObject {
  type: "http";
  description?: string;
  scheme: string;
  bearerFormat?: string;
}

interface Oauth2SecuritySchemeObject {
  type: "oauth2";
  description?: string;
  flows: OAuthFlowsObject;
}

interface OpenIdConnectSecuritySchemeObject {
  type: "openIdConnect";
  description?: string;
  openIdConnectUrl: string;
}

interface OAuthFlowsObject {
  implicit?: OAuthFlowObject;
  password?: OAuthFlowObject;
  clientCredentials?: OAuthFlowObject;
  authorizationCode?: OAuthFlowObject;
}

interface OAuthFlowObject {
  authorizationUrl?: string; // required for some
  tokenUrl?: string; // required for some
  refreshUrl?: string;
  scopes: Map<string>;
}

type SecurityRequirementObject = Map<string[]>;
