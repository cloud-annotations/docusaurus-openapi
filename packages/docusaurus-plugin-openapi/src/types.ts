import { Request } from "postman-collection";

export interface PluginOptions {
  openapiPath: string;
  routeBasePath: string;
  apiLayoutComponent: string;
  apiItemComponent: string;
  remarkPlugins: ([Function, object] | Function)[];
  rehypePlugins: string[];
  admonitions: any;
}

export interface LoadedContent {
  openapiData: any;
}

export enum Type {
  string = "string",
  number = "number",
  integer = "integer",
  boolean = "boolean",
  object = "object",
  array = "array",
}

export interface Schema {
  type?: Type;
  format?: string;
  example?: any;
  additionalProperties?: any;
  enum?: any;
  default?: any;
  deprecated?: boolean;

  properties?: Schema;
  items?: Schema;
  oneOf?: Schema;
  anyOf?: Schema;
}

export interface Primitives {
  string: { [key: string]: (schema: Schema) => any };
  number: { [key: string]: (schema: Schema) => any };
  integer: { [key: string]: (schema: Schema) => any };
  boolean: { [key: string]: (schema: Schema) => any };
  object: { [key: string]: (schema: Schema) => any };
  array: { [key: string]: (schema: Schema) => any };
}

export interface Sidebar {
  [sidebarId: string]: SidebarItem[];
}

export type SidebarItem = SidebarItemLink | SidebarItemCategory;

export interface SidebarItemLink {
  type: "link";
  href: string;
  label: string;
}

export interface SidebarItemCategory {
  type: "category";
  label: string;
  items: SidebarItem[];
  collapsed: boolean;
}

export interface Order {
  [id: string]: OrderMetadata;
}

export interface OrderMetadata {
  previous?: string;
  next?: string;
  sidebar?: string;
}

export interface OpenApiObject {
  openapi: string;
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObject;
  components?: ComponentsObject;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

export interface InfoObject {
  title: string;
  version: string;
  description?: string;
  termsOfService?: string;
  contact?: ContactObject;
  license?: LicenseObject;
}

export interface ContactObject {
  name?: string;
  url?: string;
  email?: string;
}

export interface LicenseObject {
  name: string;
  url?: string;
}

export interface ServerObject {
  url: string;
  description?: string;
  variables?: Map<ServerVariable>;
}

export interface ServerVariable {
  default: string;
  enum?: string[];
  description?: string;
}

export interface ComponentsObject {
  schemas?: Map<SchemaObject | ReferenceObject>;
  responses?: Map<ResponseObject | ReferenceObject>;
  parameters?: Map<ParameterObject | ReferenceObject>;
  examples?: Map<ExampleObject | ReferenceObject>;
  requestBodies?: Map<RequestBodyObject | ReferenceObject>;
  headers?: Map<HeaderObject | ReferenceObject>;
  securitySchemes?: Map<
    | ApiKeySecuritySchemeObject
    | HttpSecuritySchemeObject
    | Oauth2SecuritySchemeObject
    | OpenIdConnectSecuritySchemeObject
    | ReferenceObject
  >;
  links?: Map<LinkObject | ReferenceObject>;
  callbacks?: Map<CallbackObject | ReferenceObject>;
}

export interface PathsObject {
  [path: string]: PathItemObject;
}

export interface PathItemObject {
  $ref?: string;
  summary?: string;
  description?: string;
  get?: OperationObject;
  put?: OperationObject;
  post?: OperationObject;
  delete?: OperationObject;
  options?: OperationObject;
  head?: OperationObject;
  patch?: OperationObject;
  trace?: OperationObject;
  servers?: ServerObject[];
  parameters?: (ParameterObject | ReferenceObject)[];
}

export interface OperationObject {
  responses: ResponsesObject;
  tags?: string[];
  summary?: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
  operationId?: string;
  parameters?: (ParameterObject | ReferenceObject)[];
  requestBody?: RequestBodyObject | ReferenceObject;
  callbacks?: Map<CallbackObject | ReferenceObject>;
  deprecated?: boolean;
  security?: SecurityRequirementObject[];
  servers?: ServerObject[];
  postman?: Request;
}

export interface ExternalDocumentationObject {
  url: string;
  description?: string;
}

export interface ParameterObject {
  name: string;
  in: string;
  description?: string;
  required?: boolean;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  //
  style?: string;
  explode?: string;
  allowReserved?: boolean;
  schema?: SchemaObject | ReferenceObject;
  example?: any;
  examples?: Map<ExampleObject | ReferenceObject>;
  //
  content?: Map<MediaTypeObject>;
  // ignoring stylings: matrix, label, form, simple, spaceDelimited,
  // pipeDelimited and deepObject
}

export interface RequestBodyObject {
  content: Map<MediaTypeObject>;
  description?: string;
  required?: boolean;
}

export interface MediaTypeObject {
  schema?: SchemaObject | ReferenceObject;
  example?: any;
  examples?: Map<ExampleObject | ReferenceObject>;
  encoding?: Map<EncodingObject>;
}

export interface EncodingObject {
  contentType?: string;
  headers?: Map<HeaderObject | ReferenceObject>;
  style?: string;
  explode?: boolean;
  allowReserved?: boolean;
}

export interface ResponsesObject {
  [code: string]: ResponseObject | ReferenceObject;
}

export interface ResponseObject {
  description: string;
  headers?: Map<HeaderObject | ReferenceObject>;
  content?: Map<MediaTypeObject>;
  links?: Map<LinkObject | ReferenceObject>;
}

export interface CallbackObject {
  [expression: string]: PathItemObject;
}

export interface ExampleObject {
  summary?: string;
  description?: string;
  value?: any;
  externalValue?: string;
}

export interface LinkObject {
  operationRef?: string;
  operationId?: string;
  parameters?: Map<any>;
  requestBody?: any;
  description?: string;
  server?: ServerObject;
}

export interface HeaderObject {
  description?: string;
  required?: boolean;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  //
  style?: string;
  explode?: string;
  allowReserved?: boolean;
  schema?: SchemaObject | ReferenceObject;
  example?: any;
  examples?: Map<ExampleObject | ReferenceObject>;
  //
  content?: Map<MediaTypeObject>;
  // ignoring stylings: matrix, label, form, simple, spaceDelimited,
  // pipeDelimited and deepObject
}

export interface TagObject {
  name: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
}

export interface ReferenceObject {
  $ref: string;
}

// TODO: this could be expanded on.
export interface SchemaObject {
  [key: string]: any;
}

export interface ApiKeySecuritySchemeObject {
  type: string;
  description?: string;
  name: string;
  in: string;
}

export interface HttpSecuritySchemeObject {
  type: string;
  description?: string;
  scheme: string;
  bearerFormat?: string;
}

export interface Oauth2SecuritySchemeObject {
  type: string;
  description?: string;
  flows: OAuthFlowsObject;
}

export interface OpenIdConnectSecuritySchemeObject {
  type: string;
  description?: string;
  openIdConnectUrl: string;
}

export interface OAuthFlowsObject {
  implicit?: OAuthFlowObject;
  password?: OAuthFlowObject;
  clientCredentials?: OAuthFlowObject;
  authorizationCode?: OAuthFlowObject;
}

export interface OAuthFlowObject {
  authorizationUrl?: string; // required for some?
  tokenUrl?: string; // required for some?
  refreshUrl: string;
  scopes: Map<string>;
}

export interface SecurityRequirementObject {
  [name: string]: string[];
}

export interface Map<T> {
  [key: string]: T;
}

export interface ApiSection {
  title: string;
  description: string;
  items: ApiItem[];
}

export interface ApiItem extends OperationObject {
  method: string;
  path: string;
  hashId: string;
  summary: string;
  permalink: string;
  next: Page;
  previous: Page;
  jsonRequestBodyExample: string;
}

export interface Page {
  title: string;
  permalink: string;
}
