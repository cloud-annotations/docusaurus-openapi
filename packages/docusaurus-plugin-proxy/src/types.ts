/* ============================================================================
 * Copyright (c) Cloud Annotations
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import type { Configuration } from "webpack-dev-server";

export interface PluginOptions {
  proxy: Configuration["proxy"];
}
