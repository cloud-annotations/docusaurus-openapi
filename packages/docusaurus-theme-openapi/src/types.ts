/* ============================================================================
 * Copyright (c) Cloud Annotations
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

export interface ThemeConfig {
  api?: {
    proxy?: string;
    authPersistance?: false | "localStorage" | "sessionStorage";
    serverVariablesPersistance?: false | "localStorage" | "sessionStorage";
  };
}
