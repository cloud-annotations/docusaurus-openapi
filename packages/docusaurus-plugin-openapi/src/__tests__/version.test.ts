/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {docsVersion} from '../version';
import {PathOptions} from '../types';
import fs from 'fs-extra';
import {
  getVersionedDocsDir,
  getVersionsJSONFile,
  getVersionedSidebarsDir,
} from '../env';

const fixtureDir = path.join(__dirname, '__fixtures__');

describe('docsVersion', () => {
  const simpleSiteDir = path.join(fixtureDir, 'simple-site');
  const versionedSiteDir = path.join(fixtureDir, 'versioned-site');
  const DEFAULT_OPTIONS: PathOptions = {
    path: 'docs',
    sidebarPath: '',
  };

  test('no version tag provided', () => {
    expect(() =>
      docsVersion(null, simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"No version tag specified!. Pass the version you wish to create as an argument. Ex: 1.0.0"`,
    );
    expect(() =>
      docsVersion(undefined, simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"No version tag specified!. Pass the version you wish to create as an argument. Ex: 1.0.0"`,
    );
    expect(() =>
      docsVersion('', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"No version tag specified!. Pass the version you wish to create as an argument. Ex: 1.0.0"`,
    );
  });

  test('version tag should not have slash', () => {
    expect(() =>
      docsVersion('foo/bar', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Do not include slash (/) or (\\\\). Try something like: 1.0.0"`,
    );
    expect(() =>
      docsVersion('foo\\bar', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Do not include slash (/) or (\\\\). Try something like: 1.0.0"`,
    );
  });

  test('version tag should not be too long', () => {
    expect(() =>
      docsVersion('a'.repeat(255), simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Length must <= 32 characters. Try something like: 1.0.0"`,
    );
  });

  test('version tag should not be a dot or two dots', () => {
    expect(() =>
      docsVersion('..', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Do not name your version \\".\\" or \\"..\\". Try something like: 1.0.0"`,
    );
    expect(() =>
      docsVersion('.', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Do not name your version \\".\\" or \\"..\\". Try something like: 1.0.0"`,
    );
  });

  test('version tag should be a valid pathname', () => {
    expect(() =>
      docsVersion('<foo|bar>', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Please ensure its a valid pathname too. Try something like: 1.0.0"`,
    );
    expect(() =>
      docsVersion('foo\x00bar', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Please ensure its a valid pathname too. Try something like: 1.0.0"`,
    );
    expect(() =>
      docsVersion('foo:bar', simpleSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid version tag specified! Please ensure its a valid pathname too. Try something like: 1.0.0"`,
    );
  });

  test('version tag already exist', () => {
    expect(() =>
      docsVersion('1.0.0', versionedSiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"This version already exists!. Use a version tag that does not already exist."`,
    );
  });

  test('no docs file to version', () => {
    const emptySiteDir = path.join(fixtureDir, 'empty-site');
    expect(() =>
      docsVersion('1.0.0', emptySiteDir, DEFAULT_OPTIONS),
    ).toThrowErrorMatchingInlineSnapshot(`"There is no docs to version !"`);
  });

  test('first time versioning', () => {
    const copyMock = jest.spyOn(fs, 'copySync').mockImplementation();
    const ensureMock = jest.spyOn(fs, 'ensureDirSync').mockImplementation();
    const writeMock = jest.spyOn(fs, 'writeFileSync');
    let versionedSidebar;
    let versionedSidebarPath;
    writeMock.mockImplementationOnce((filepath, content) => {
      versionedSidebarPath = filepath;
      versionedSidebar = JSON.parse(content);
    });
    let versionsPath;
    let versions;
    writeMock.mockImplementationOnce((filepath, content) => {
      versionsPath = filepath;
      versions = JSON.parse(content);
    });
    const consoleMock = jest.spyOn(console, 'log').mockImplementation();
    const options = {
      path: 'docs',
      sidebarPath: path.join(simpleSiteDir, 'sidebars.json'),
    };
    docsVersion('1.0.0', simpleSiteDir, options);
    expect(copyMock).toHaveBeenCalledWith(
      path.join(simpleSiteDir, options.path),
      path.join(getVersionedDocsDir(simpleSiteDir), 'version-1.0.0'),
    );
    expect(versionedSidebar).toMatchSnapshot();
    expect(versionedSidebarPath).toEqual(
      path.join(
        getVersionedSidebarsDir(simpleSiteDir),
        'version-1.0.0-sidebars.json',
      ),
    );
    expect(versionsPath).toEqual(getVersionsJSONFile(simpleSiteDir));
    expect(versions).toEqual(['1.0.0']);
    expect(consoleMock).toHaveBeenCalledWith('Version 1.0.0 created!');

    copyMock.mockRestore();
    writeMock.mockRestore();
    consoleMock.mockRestore();
    ensureMock.mockRestore();
  });

  test('not the first time versioning', () => {
    const copyMock = jest.spyOn(fs, 'copySync').mockImplementation();
    const ensureMock = jest.spyOn(fs, 'ensureDirSync').mockImplementation();
    const writeMock = jest.spyOn(fs, 'writeFileSync');
    let versionedSidebar;
    let versionedSidebarPath;
    writeMock.mockImplementationOnce((filepath, content) => {
      versionedSidebarPath = filepath;
      versionedSidebar = JSON.parse(content);
    });
    let versionsPath;
    let versions;
    writeMock.mockImplementationOnce((filepath, content) => {
      versionsPath = filepath;
      versions = JSON.parse(content);
    });
    const consoleMock = jest.spyOn(console, 'log').mockImplementation();
    const options = {
      path: 'docs',
      sidebarPath: path.join(versionedSiteDir, 'sidebars.json'),
    };
    docsVersion('2.0.0', versionedSiteDir, options);
    expect(copyMock).toHaveBeenCalledWith(
      path.join(versionedSiteDir, options.path),
      path.join(getVersionedDocsDir(versionedSiteDir), 'version-2.0.0'),
    );
    expect(versionedSidebar).toMatchSnapshot();
    expect(versionedSidebarPath).toEqual(
      path.join(
        getVersionedSidebarsDir(versionedSiteDir),
        'version-2.0.0-sidebars.json',
      ),
    );
    expect(versionsPath).toEqual(getVersionsJSONFile(versionedSiteDir));
    expect(versions).toEqual(['2.0.0', '1.0.1', '1.0.0']);
    expect(consoleMock).toHaveBeenCalledWith('Version 2.0.0 created!');

    copyMock.mockRestore();
    writeMock.mockRestore();
    consoleMock.mockRestore();
    ensureMock.mockRestore();
  });
});
