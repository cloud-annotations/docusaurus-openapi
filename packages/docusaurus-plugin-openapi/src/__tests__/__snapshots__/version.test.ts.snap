// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docsVersion first time versioning 1`] = `
Object {
  "version-1.0.0/docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "collapsed": true,
          "items": Array [
            Object {
              "id": "version-1.0.0/foo/bar",
              "type": "doc",
            },
            Object {
              "id": "version-1.0.0/foo/baz",
              "type": "doc",
            },
          ],
          "label": "foo",
          "type": "category",
        },
        Object {
          "href": "https://github.com",
          "label": "Github",
          "type": "link",
        },
        Object {
          "id": "version-1.0.0/hello",
          "type": "ref",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "version-1.0.0/hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "type": "category",
    },
  ],
}
`;

exports[`docsVersion not the first time versioning 1`] = `
Object {
  "version-2.0.0/docs": Array [
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "version-2.0.0/foo/bar",
          "type": "doc",
        },
      ],
      "label": "Test",
      "type": "category",
    },
    Object {
      "collapsed": true,
      "items": Array [
        Object {
          "id": "version-2.0.0/hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "type": "category",
    },
  ],
}
`;
